program/src/processor.rs

```diff
@@ -1,25 +1,169 @@
+use solana_program::program::invoke_signed;
+use solana_program::program_error::ProgramError;
+use solana_program::program_pack::Pack;
+use solana_program::rent::Rent;
+use solana_program::sysvar::Sysvar;
+use solana_program::{msg, system_instruction};
+use spl_token_2022::instruction::initialize_mint2;
 use {
     crate::instruction::TokenWrapInstruction,
     solana_program::{account_info::AccountInfo, entrypoint::ProgramResult, pubkey::Pubkey},
-    spl_token_2022::instruction::decode_instruction_type,
 };
 
+/// Processes [`CreateMint`](enum.TokenWrapInstruction.html) instruction.
+pub fn process_create_mint(
+    program_id: &Pubkey,
+    accounts: &[AccountInfo],
+    idempotent: bool,
+) -> ProgramResult {
+    let account_info_iter = &mut accounts.iter();
+
+    let wrapped_mint_account = next_account_info(account_info_iter)?;
+    let wrapped_backpointer_account = next_account_info(account_info_iter)?;
+    let unwrapped_mint_account = next_account_info(account_info_iter)?;
+    let _system_program_account = next_account_info(account_info_iter)?;
+    let wrapped_token_program_account = next_account_info(account_info_iter)?;
+
+    // Idempotency checks
+
+    if wrapped_mint_account.data_len() > 0 || wrapped_backpointer_account.data_len() > 0 {
+        msg!("Wrapped mint or backpointer account already initialized");
+        return if !idempotent {
+            Err(ProgramError::AccountAlreadyInitialized)
+        } else {
+            Ok(())
+        };
+    }
+
+    // Initialize wrapped mint PDA
+
+    let (wrapped_mint_address, bump) = get_wrapped_mint_address_with_seed(
+        unwrapped_mint_account.key,
+        wrapped_token_program_account.key,
+    );
+    let bump_seed = [bump];
+    let signer_seeds = _get_wrapped_mint_signer_seeds(
+        unwrapped_mint_account.key,
+        wrapped_token_program_account.key,
+        &bump_seed,
+    );
+    let space = spl_token_2022::state::Mint::get_packed_len();
+
+    let rent = Rent::get()?;
+    let mint_rent_required = rent.minimum_balance(space);
+    if wrapped_mint_account.lamports() < mint_rent_required {
+        msg!(
+            "Error: wrapped_mint_account requires pre-funding of {} lamports",
+            mint_rent_required
+        );
+        return Err(ProgramError::InsufficientFunds);
+    }
+
+    // Initialize the wrapped mint
+
+    invoke_signed(
+        &system_instruction::allocate(&wrapped_mint_address, space as u64),
+        &[wrapped_mint_account.clone()],
+        &[&signer_seeds],
+    )?;
+    invoke_signed(
+        &system_instruction::assign(&wrapped_mint_address, wrapped_token_program_account.key),
+        &[wrapped_mint_account.clone()],
+        &[&signer_seeds],
+    )?;
+
+    // New wrapped mint matches decimals & freeze authority of unwrapped mint
+    let unwrapped_mint_data = unwrapped_mint_account.try_borrow_data()?;
+    let unpacked_unwrapped_mint = spl_token_2022::state::Mint::unpack(&unwrapped_mint_data)?;
```

Comment:
Since the unwrapped mint might be a token-2022 mint with extensions, `unpack` will fail -- you'll need to use `StateWithExtensions::unpack`, ie https://github.com/solana-program/token-2022/blob/f69cc989c2a3f2657a6ad6041018468807223f08/program/src/extension/mod.rs#L1655

Bonus if you use the Pod variants, since they're zero-copy :grin: